syft/frameworks/torch/tensors/interpreters/precision.py

```diff
@@ -223,47 +225,100 @@ def mul(self, other):
+                "mul", response, wrap_type=type(self), wrap_args=self.get_class_attributes()
+            )
+
+            response = response.truncate(self.precision_fractional)
+            response %= self.field  # Wrap around the field
+
+        elif isinstance(other, FixedPrecisionTensor):
+            assert (
+                self.precision_fractional == other.precision_fractional
+            ), "In mul, all args should have the same precision_fractional"
 
-        else:
             # Replace all syft tensor with their child attribute
             new_self, new_other, _ = syft.frameworks.torch.hook_args.hook_method_args(
                 "mul", self, other, None
             )
 
-        # Send it to the appropriate class and get the response
-        response = getattr(new_self, "mul")(new_other)
+            # To avoid overflow due to multiplication of small negative numbers
+            # we take absolute value of the operands
 
-        # Put back SyftTensor on the tensors found in the response
-        response = syft.frameworks.torch.hook_args.hook_response(
-            "mul", response, wrap_type=type(self), wrap_args=self.get_class_attributes()
-        )
+            # sgn_self is 1 when new_self is positive else it's 0
+            sgn_self = (new_self < self.field // 2).long()
+            pos_self = new_self * sgn_self
+            neg_self = (self.field - new_self) * (1 - sgn_self)
+            new_self = neg_self + pos_self
 
-        response %= self.field  # Wrap around the field
-        if not isinstance(other, int):
-            response = response.truncate(self.precision_fractional)
+            # sgn_other is 1 when new_other is positive else it's 0
+            sgn_other = (new_other < self.field // 2).long()
+            pos_other = new_other * sgn_other
+            neg_other = (self.field - new_other) * (1 - sgn_other)
+            new_other = neg_other + pos_other
+
+            # If both have the same sign, sgn is 1 else it's 0
+            sgn = 1 - (sgn_self - sgn_other) ** 2
+
+            # Send it to the appropriate class and get the response
+            response = getattr(new_self, "mul")(new_other)
+
+            # Put back SyftTensor on the tensors found in the response
+            response = syft.frameworks.torch.hook_args.hook_response(
+                "mul", response, wrap_type=type(self), wrap_args=self.get_class_attributes()
+            )
+
+            # With this sign trick, we don't need to consider negative numbers when truncating
+            response = response.truncate(self.precision_fractional, check_sign=False)
+            response %= self.field  # Wrap around the field
+
+            # Give back its sign to response
+            pos_res = response * sgn
+            neg_res = (self.field - response) * (1 - sgn)
+            response = neg_res + pos_res
 
         return response
 
     __mul__ = mul
+    __imul__ = mul
```

Comment:
inplace operation can't be equivalent to out of place ops, maybe remove this line or make sure you don't create inconsistencies

Reply:
Is it more correct to overload __imul__ as I did it now?